### PR TITLE
Qt6 build checks fixes

### PR DESCRIPTION
**Pull request by daschuer**
After switching to Qt6 clazy and clang-tidy are complaining some issues. 
This branch fixes these issues, but keeps the CI on Qt 5. 

This is a prerequisite for a Qt6 switch, but there is some more missing. 
These changes have been tested also with an experimental Qt6 CI here: 

https://github.com/daschuer/mixxx/actions/runs/5435953432

The remaining warning is a false positive fixed in the 2.4 branch: 
https://github.com/mixxxdj/mixxx/pull/11689

### CODE REVIEW

File: src/util/make_const_iterator.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+template<typename C, typename I>
+typename C::const_iterator make_const_iterator(const C&, const I& it) {
+    return static_cast<typename C::const_iterator>(it);
+}
+
+template<typename C, typename I>
+typename C::iterator make_iterator(C* pContainer, I it) {
+    return std::next(pContainer->begin(), std::distance(pContainer->cbegin(), it));
+}
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/make_const_iterator.h:13]: This makes an iterator from a const iterator. We need that function as a compatibility layer for the Qt5 API. 
A direct cast is not possible.

[the remark this answers is not in the document]
Comment (uklotzde) [reply src/util/make_const_iterator.h:13]: Answering this question with a PR comment is pointless. Please alter the code to reveal the purpose, e.g. `make_non_const_iterator(...)`.

Btw, renaming this function reveals that the file name is inappropriate and confusing now.

[the remark this answers is not in the document]
Comment (uklotzde) [reply src/util/make_const_iterator.h:13]: Well, the file name still applies if these are supposed to be utility function around *const iterators*. Could be left as is.

[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/util/make_const_iterator.h:13]: I see, but wouldn't it not make more sense instead to just change the callsites to obtain a mutable iterator from the start instead of reconstructing it?

[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/make_const_iterator.h:13]: This is used as a wrapper to make Qt5 functions behave like Qt6 functions ina a way that we can easily drop it in a future version. 

Your suggestion will lead into the other way round which is harder to drop later. 

I consider the perfomance penalty as neglectible as we aim for a Qt6 release in all our targets anyway.

[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/util/make_const_iterator.h:17]: I see, I'll need some time to take a closer look at the Qt6 API to properly review this.

File: src/controllers/scripting/controllerscriptenginebase.h [hunk 2]
```diff
@@ -25,7 +25,7 @@ class ControllerScriptEngineBase : public QObject {
 
     virtual bool initialize();
 
-    bool executeFunction(QJSValue functionObject, const QJSValueList& arguments = {});
+    bool executeFunction(QJSValue* pFunctionObject, const QJSValueList& arguments = {});
```
Comment (Swiftb0y) [inline src/controllers/scripting/controllerscriptenginebase.h:28]: why the change to pointer? Don't see how that was necessary.

File: src/util/make_const_iterator.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+template<typename C, typename I, class T>
+typename C::const_iterator constInsert(C* pContainer, I before, T t) {
+#if defined(QT_VERSION) && QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+    return make_const_iterator(*pContainer, pContainer->insert(before, std::move(t)));
+#else
+    return make_const_iterator(*pContainer,
+            pContainer->insert(make_iterator(pContainer, before), t));
+#endif
+}
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/make_const_iterator.h:54]: Clang-tidy requires an explicit cast when copying to a const iterator. The current signature has this cast build in to follow this demand. This way it is a no-brainer for the user.

[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/make_const_iterator.h:54]: please not that make_const_iterator() is a nop (unlike make_inerator())

File: src/util/rangelist.cpp [hunk 4]
```diff
@@ -40,8 +41,9 @@ QList<int> parseRangeList(const QString& input) {
     }
 
     std::sort(intList.begin(), intList.end());
-    const auto end = std::unique(intList.begin(), intList.end());
-    intList.erase(end, intList.end());
+    const auto constLast = make_const_iterator(
+            intList, std::unique(intList.begin(), intList.end()));
+    constErase(&intList, constLast, intList.constEnd());
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/rangelist.cpp:46]: Yes, Qt has followed with Qt6.

File: src/controllers/scripting/controllerscriptenginebase.h (hunk 2 above)
Comment (daschuer) [reply src/controllers/scripting/controllerscriptenginebase.h:28]: Because clang-tidy did not like the use of the copy by value. Maybe some interns have changed in Qt6 that makes the copy expensive?

File: src/util/rangelist.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/util/rangelist.cpp:46]: Right, I'm saying we could avoid the workaround here by refactoring the function to use a `std::vector` instead of QList. Then its unaffected by the version transition.

File: src/controllers/scripting/controllerscriptenginebase.h (hunk 2 above)
Comment (Swiftb0y) [reply src/controllers/scripting/controllerscriptenginebase.h:28]: Pretty sure its still cheap. Maybe a clang-tidy bug? From https://doc.qt.io/qt-6/qjsvalue.html#QJSValue-1:

> Note that if other is an object (i.e., isObject() would return true), then only a reference to the underlying object is copied into the new script value (i.e., the object itself is not copied).

File: src/util/make_const_iterator.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/util/make_const_iterator.h:54]: Right, but we're downgrading capabilities of the the returned iterator for no reason. If the user actually wanted a non-const iterator, they'd have to call `make_iterator` again on the return value. If we leave the decision over the constness to the caller we can simplify the code in the callee and give more options to the caller.

**Comment by Swiftb0y** [comment]
To be super honest, I don't expect the make_const_iterator code to stick around very long anyways, since after a the Qt6 update, we can remove the ifdefs and then manually inline those utility functions, likely simplifying their callsites in the process. So I'm not super concerned about getting this code in shape. Just wanted to voice my opinions.

File: src/controllers/scripting/controllerscriptenginebase.h (hunk 2 above)
Comment (daschuer) [reply src/controllers/scripting/controllerscriptenginebase.h:28]: In any case passing the pointer is one of the cheapest option and it is IMHO better than maintaining a clang-tidy exception.

File: src/util/rangelist.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/rangelist.cpp:46]: Yes sure. That will start a bigger refactoring.
QList is not like std::vector we should consider such changes along the best container for a specific task. Doing it alone to avoid a temporary workaround is not worth the work IMHO.

File: src/util/make_const_iterator.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/make_const_iterator.h:54]: The reason is to please clang-tidy. 

The wrapper is not designed for the mentioned  case. It's for the const case, indicated by it's name. 

If we really have such case during the transition of qt5 to qt6 we can invent another wrapoer for this. For now, I like the convenience to not see the cast in the using code.

File: src/util/rangelist.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/util/rangelist.cpp:46]: > QList is not like std::vector we should consider such changes along the best container for a specific task.

Unrelated to the discussion at hand but please elaborate. The only difference I know of is the implicit sharing COW sharing.

[the remark this answers is not in the document]
Comment (daschuer) [reply src/util/rangelist.cpp:46]: Yes that is one point. The other is that QList is like a vector of pointers. Every small type is stored directly. Bigger types are allocated on the heap.

**Comment by daschuer** [comment]
Is anything left to do here?

**Comment by Swiftb0y** [comment]
well, I stand by my previous comments, but I won't object if someone else wants to LGTM this PR.

**Comment by daschuer** [comment]
> well, I stand by my previous comments, but I won't object if someone else wants to LGTM this PR.

What exactly is your demand in terms of a TODO?

File: src/util/rangelist.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoergAtGithub) [reply src/util/rangelist.cpp:46]: @daschuer Did you tried to use QVector or std::vector instead?

**Comment by Swiftb0y** [comment]
> What exactly is your demand in terms of a TODO?

Not demands, just suggestions that IMO improve the code. I'll summarize:

1. move `make_const_iterator` to the callsites of `constErase` and `constInsert` and only use where required. Carefully examine `clazy-strict-iterators` issues in the process as I fear the utility functions in `make_const_iterator.h` might mask the issue highlighted by the warning.
2. refactor the algorithm in `rangelist.h` to use `std:vector` to avoid having to resort to `constErase` there.
3. determine why passing `QJSValue` by value is no longer cheap. pass using const-ref instead and potentially copy in the body instead of using a mutable pointer.

**Comment by daschuer** [comment]
1. I don't like the idea, because it clutters IMHO the caller side, is against the DRY rule. Now we use the constErase wrapper to use only const iterators. That creates easy to read code. All boilerplate is hidden into the wrapper. 

2. That's a good idea, but a big refactoring, out of scope of this PR. My intention is to switch main to Qt6 quickly. It is IMHO mandatory for going to beta in 4 weeks. 

3. Got it: QJSValue has been changed to a quint64 from quintptr. The later can be always copied cheap the new one only with a 64 bit architecture.

**Comment by Swiftb0y** [comment]
> I don't like the idea, because it clutters IMHO the caller side, is against the DRY rule. Now we use the constErase wrapper to use only const iterators. That creates easy to read code. All boilerplate is hidden into the wrapper.

I disagree, but that depends on specifics I'm too busy to work out right now. I'm fine with ignoring this for now in order to get this merged. 

> That's a good idea, but a big refactoring, out of scope of this PR. My intention is to switch main to Qt6 quickly. It is IMHO mandatory for going to beta in 4 weeks.

I don't think that refactoring is big to be honest, the code can largely stay as is (IIRC; after all I implemented it, judging from the bad naming :upside_down_face:). The only thing I fear is that it carries a little overhead at the Qt-API boundary because the `std::vector` would need to be converted to a QList. 

> Got it: QJSValue has been changed to a quint64 from quintptr. The later can be always copied cheap the new one only with a 64 bit architecture. qt/qtdeclarative@748411f

So whats the next step here? Giving up value-semantics solely because it's a little more expensive to copy on 32-bit-only builds is not the best option IMO.

**Comment by Swiftb0y** [comment]
> I think we should not go for s conversion if containers. With refactoring I had considered to use std::vector in all related code. That would be a good thing, the container conversion just to avoid a temporary Qt5 wrapoer is not a good trade of.

I'd normally agree, but the code in question is cold and the container is small (storing less than `kNumHotcues` `int`s). 

> This makes sense in a static code analyzis view, replacing all copy by value > void*

I don't understand what you're saying here. How does that transformation make sense for static code analyzers?

**Comment by daschuer** [comment]
> I'd normally agree, but the code in question is cold and the container is small (storing less than kNumHotcues ints).

Yes, but it's a bad idea to do a refactoring half way. I consider to do the full swap, but not in this PR. 

>>  This makes sense in a static code analyzis view, replacing all copy by value > void*

> I don't understand what you're saying here. How does that transformation make sense for static code analyzers?

The analyzers have a simple rule to not copy values > void*
The have no exception for our case.  The solution is to std::move the values if they are kept or use a const reference or a pointer otherwise. That it.

**Comment by Swiftb0y** [comment]
> The analyzers have a simple rule to not copy values > void*
The have no exception for our case. The solution is to std::move the values if they are kept or use a const reference or a pointer otherwise. That it.

Ah, I have understood now. Well can't we tell the analyzer to assume we're on x86-64? we're only providing 64-bit builds anyway, right?

**Comment by daschuer** [comment]
> Well can't we tell the analyzer to assume we're on x86-64? 

Not sure. We may investigate. This will probably leads to an individual solution for each analyzer almost like an individual disable statement for each analyzer. I prefer a fix in the C++ code for less maintenance. 

> We're only providing 64-bit builds anyway, right?

Yes.

**Comment by daschuer** [comment]
> refactor the algorithm in rangelist.h to use std:vector to avoid having to resort to constErase there.

I have now added a refactoring commit that works around constErase by adding the elements at without duplication. 

@Swiftb0y Can you have another look and merge if it is OK for you? 

This way we can proceed switching main to QT6. A prerequisite for the upcoming 2.5-beta IMHO.

**Comment by daschuer** [comment]
@Swiftb0y It would be great to merge this before https://github.com/mixxxdj/mixxx/pull/11839 to avoid a locking situation. 
Can you have another look? Thank you.

File: src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp [hunk 5]
```diff
@@ -192,8 +192,8 @@ bool ControllerScriptEngineLegacy::handleIncomingData(const QByteArray& data) {
             static_cast<uint>(data.size()),
     };
 
-    for (const QJSValue& function : std::as_const(m_incomingDataFunctions)) {
-        ControllerScriptEngineBase::executeFunction(function, args);
+    for (auto&& function : m_incomingDataFunctions) {
```
Comment (Swiftb0y) [inline src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp:195]: why the `&&`? does/could that cause a move out of the container?

File: src/util/rangelist.cpp [hunk 6]
```diff
@@ -35,14 +36,13 @@ QList<int> parseRangeList(const QString& input) {
             endIndex = startIndex;
         }
         for (int currentIndex = startIndex; currentIndex <= endIndex; currentIndex++) {
-            intList.append(currentIndex);
+            // Add values sorted and skip duplicates
+            auto insertPos = std::lower_bound(intList.begin(), intList.end(), currentIndex);
+            if (insertPos == intList.end() || *insertPos != currentIndex) {
+                intList.insert(insertPos, currentIndex);
+            }
```
Comment (Swiftb0y) [inline src/util/rangelist.cpp:43]: technically quite more inefficient because the insert is not constant time but ill let it slide...

File: src/util/rangelist.cpp [hunk 7]
```diff
@@ -1,10 +1,11 @@
 #include "rangelist.h"
 
-#include <util/assert.h>
-
 #include <QRegularExpression>
 #include <algorithm>
 
+#include "util/assert.h"
+#include "util/make_const_iterator.h"
```
Comment (Swiftb0y) [inline src/util/rangelist.cpp:7]: unnecessary include after latest change
```suggestion
```

File: src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp (hunk 5 above)
Comment (daschuer) [reply src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp:195]: No, in this case this is the "universal reference"  or a "forwarding reference" that reliable hat the type of *iterator. 
https://stackoverflow.com/questions/13230480/what-does-auto-tell-us
This comes from the pattern to use `const auto&` and `auto&&` as an universal rule of thump that is always correct. 
In this case we have a QList where *iterator is QJSValue&
https://en.cppreference.com/w/cpp/language/range-for dos not even mention the `auto&` case
I can change that, but how do we want to continue if Mixxx in general?

Comment (Swiftb0y) [reply src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp:195]: > how do we want to continue in Mixxx in general?

I'm not sure, if in doubt, consult the cpp core guidelines. The problem right now for me is that the topic is highly intertwined with overload resolution and value categories so I've not managed to wrap my head around the topic yet. 

keep it as it for now. I'm sure there won't be any move since the value is named.

**Review by Swiftb0y — APPROVED** [review]
LGTM. Qt6 here we come
